api/v1/loadtest_types.go

```diff
@@ -208,6 +208,24 @@ type LoadTestSpec struct {
 	Scenarios []Scenario `json:"scenarios,omitempty"`
 }
 
+// LoadTestMissing categorize missing components base on their roles at
+// the moment.
+type LoadTestMissing struct {
+	// Driver is the component that orchestrates the test. It may be
+	// unspecified, allowing the system to choose the appropriate driver.
+	// +optional
+	Driver *Driver `json:"driver,omitempty"`
+
+	// Servers are a list of components that receive traffic from
+	// clients.
+	// +optional
+	Servers []Server `json:"servers,omitempty"`
+
+	// Clients are a list of components that send traffic to servers.
+	// +optional
+	Clients []Client `json:"clients,omitempty"`
+}
+
```

Comment:
This is a great struct. I think we should consider moving this out of api/v1/loadtest_types.go for 1 reason. Everything in api/ is considered public API, and Kubebuilder generates a lot of code for it. For instance, it creates `DeepCopy` methods. Since this type is only meant for internal use, I think it would be better to move it into controllers/loadtest_controller.go as an unexported type. Then, we can remove the `//+optional` markers and no code will be generated for it. If we leave it here, we will need to generate that code as part of this PR. What are your thoughts?

Reply:
I agree, I don't think LoadTestMissing will be used externally, I put it there just because I though all the types should go to that file, turned out to be not. :) I will move it to loadtest_controller.go, location-wise, is there a fashion to follow? I saw you put the LoadTestReconciler right on top, should my struct be closer to my func or it should be declared on the top just after LoadTestReconciler struct?

Reply:
If you do move it, be sure to run `git restore --source=upstream/master api/v1/zz_generated.deepcopy.go` to restore the previous version of the generated code.

Reply:
Also do I still need to have the comments in there?

Reply:
@wanlin31 I like to keep things close to where they are used. Since it's not exported, I would not put it above the LoadTestReconciler. -- Be sure to run `git restore --source upstream/master api/v1/zz_generated.deepcopy.go`

Reply:
@wanlin31 you do not need the `// +optional` markers, but I'd leave the other comments. @paulosjca had a discussion about this recently with my pkg/defaults code. Even though comments are not required for unexported types, it can help other people understand it. It doesn't ever hurt to leave it :)

Reply:
:) Absolutely, I add a little more explanation there just to help people understand what LoadTestMissing is.